### PR TITLE
Stop the shared RMSNorm kernel from recompiling per norm instance

### PR DESCRIPTION
**Pull request by danielhanchen**
### The problem

`Gemma4RMSNorm.forward` is a single compiled code object shared by every RMSNorm instance in the model. `gemma-4-E2B` has 504 of them, so that one frame's Dynamo cache holds the **product** of every axis its guards can see, not the sum.

Measured with `TORCH_LOGS=recompiles,recompiles_verbose` on `unsloth/gemma-4-E2B-it` with `UNSLOTH_FORCE_FLOAT32=1`. Of 25 recompiles across the whole model, 17 are this one frame; every other frame recompiles 1 to 3 times.

| axis | width | guard |
| --- | --- | --- |
| `self._parameters['weight']` size | 8 | `size mismatch at index 0. expected 64, actual 768` |
| input rank | 2 | `rank mismatch. expected 4, actual 3` (residual norms vs q/k norms) |
| `self.with_scale` | 2 | constant baked into the graph |
| input dtype | 3 | intrinsic to the fp32 patch |
| `GLOBAL_STATE: grad_mode` | 2 | gradient checkpointing packs under `no_grad`, recomputes under `enable_grad` |
| `requires_grad` | 2 | same |

That is roughly 384 entries for one frame before per-dimension specialisation from varying image sizes. On a T4 it reached the recompile limit and raised `FailOnRecompileLimitHit`, which then took activation checkpointing down with it.

`dynamic = True` cannot fix this. `torch/_dynamo/utils.py` forces static shapes whenever `type(t) is torch.nn.Parameter`, so reading `self.weight` inside the compiled region pins the hidden size regardless.

### The fix

Compile a pure tensor kernel instead of a bound method, so `self` never reaches the guards:

- the weight goes in as a plain `Tensor` view, which takes normal dynamic-shape treatment while autograd still reaches the `Parameter`. This is local to the norm kernels rather than relaxing `force_parameter_static_shapes` globally, which would cost Inductor its static GEMM specialisations everywhere.
- the input is flattened to rank 2. Normalisation only ever touches the last dimension, so this is a view for contiguous inputs and folds away when the kernel is inlined.
- scaled and unscaled get separate kernels, which removes the `with_scale` axis and the `KeyError` on `self._parameters['weight']` for the unscaled instances.

What remains is the irreducible set: dtype, grad mode and `requires_grad`.

Gemma3 and Qwen3-MoE norms have the same shape of problem and get the same treatment.

`publish_to_modeling_module` is needed because `create_standalone_class` copies the live `forward` source into `unsloth_compiled_cache` and resolves free names against the modeling module; without it the regenerated `unsloth_compiled_module_gemma4.py` raises `NameError` on import.

### Results

| config | RMSNorm recompiles | max cache entries per frame | at `recompile_limit=8` |
| --- | --- | --- | --- |
| before | 17, saturates at step 1 | 17 | `FailOnRecompileLimitHit` |
| `force_parameter_static_shapes=False` alone | 12 | 12 | still fails |
| this PR | 5 | 5 | passes |

### Correctness

Forward and backward are **bitwise identical** to the old body across every shape, dtype and `with_scale` combination tested, and gradients still reach the `Parameter` through the view.

One measurement note worth recording: comparing the old eager body against the new compiled kernel shows deltas up to 4e-3 in bf16. That is Inductor's float reassociation, not this change. The old path was compiled too, so it is not a difference in behaviour. Isolating the refactor by running both paths eagerly gives 0 mismatches in 21 of 21 cases.

### Tests

`tests/test_rmsnorm_recompile_guards.py`, 24 tests. The pair that matters: at `recompile_limit=8` the old bound-method form raises `FailOnRecompileLimitHit` over a realistic spread of call shapes, and the new kernels survive the identical spread.

Full zoo suite: `3 failed, 3856 passed, 162 skipped`. The 3 failures are `test_grpo_packed_verify_raw_logits` and are pre-existing suite pollution, not from this change. A pristine `origin/main` control fails exactly the same 3 in a full run (`3 failed, 3832 passed`, the 24 difference being the tests added here) while passing them in isolation on both branches.

### Caveat

The causal chain from recompile pressure to the checkpoint assert is proven at microbench scale with real `torch.compile` under real `checkpoint(use_reentrant=False)`, and the recompile reduction is measured on the real model. It has not yet been reproduced end to end on real Gemma4 on a T4, which saturates at 17 recompiles on larger cards and so never approaches the limit there.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Negative result on the motivating case, recording it here rather than leaving the PR body's caveat to imply otherwise.

I ran `Gemma4_(E2B)-Vision.ipynb` unedited on a Colab T4 with this branch pinned. It still fails, same signature:

```
FailOnRecompileLimitHit: Hard failure due to fullgraph=True
Dynamo attempted to recompile the code object too many times, exceeding the
recompile_limit cache size limit (currently set to 1024)
```

The branch was genuinely live, so this is not the ambiguity that made an earlier run hard to read. The artifact carries the commit:

```
NBREGRESS_NOTE unsloth_zoo commit 812a5141b69c rev pr/rmsnorm-recompile-guards
```

And the frame that exhausts the budget is the kernel this PR introduces:

```
transformers/models/gemma4/modeling_gemma4.py in forward
-> 1399   hidden_states = self.post_per_layer_input_norm(hidden_states)
unsloth_zoo/temporary_patches/gemma4_float32.py in forward
   337    hidden_states_2d, shape = flatten_for_elementwise_norm(hidden_states)
-> 339    normed = _gemma4_rms_norm_scaled(
unsloth_zoo/temporary_patches/gemma4_float32.py in _gemma4_rms_norm_scaled
-> 303    normed_fp32 = x_fp32 * torch.pow(variance + eps, -0.5)
```

reached through `torch/utils/checkpoint.py` `pack_hook`, i.e. during activation-checkpoint recompute.

So the change moves the compiled unit without bounding the cache. Removing the parameter-width, rank and `with_scale` axes is real and measured, but it is not what was consuming the 1024 entries, and the B200 repro could not have shown this: it saturates at 17 recompiles and never approaches the limit, so the remaining axes looked irreducible there.

What still stands, independent of the above:

- forward and backward are bitwise identical to the old body across every shape, dtype and `with_scale` combination tested
- the frame goes from 17 cache entries to 5 on the measured model
- at `recompile_limit=8` the old bound-method form raises `FailOnRecompileLimitHit` over a realistic spread of call shapes and the new kernels do not

What is now disproven is the motivating claim, that this closes the T4 failure. It does not.

Leaving this open as a guard-count improvement rather than a fix for that notebook, and following up with `TORCH_LOGS=recompiles` on the T4 itself to find what actually produces hundreds of entries under checkpoint recompute. Happy to close it instead if the reduction is not judged worth carrying on its own.

**Comment by danielhanchen** [comment]
Follow-up measurements, since I said I would go and find what actually consumes the 1024 entries.

Short version: it is not this kernel's guards. I can bound them, and the ceiling is nowhere near the limit.

Each row resets Dynamo, raises `recompile_limit` to 4096 so nothing is truncated, and counts `unique_graphs` on a B200:

| what was varied | calls | unique graphs |
| --- | ---: | ---: |
| leading dim N, 200 distinct sizes | 200 | **1** |
| N x 8 norm widths | 400 | **1** |
| N x widths x 3 dtypes x grad on/off | 2496 | **6** |
| rank-3 contiguous, 60 distinct (B, S) | 60 | **2** |
| rank-4 contiguous, 60 distinct (B, H, S) | 60 | **3** |
| mixed rank 3 and 4 alternating | 60 | **4** |
| 60 distinct Parameters, same width | 60 | **1** |

Two things worth reading off that.

The width axis really is gone: 8 distinct norm widths give 1 graph, which is what this PR set out to do. And the last row rules out a hazard I went looking for in my own change, that `unwrap_norm_weight` hands the kernel a view and each of the model's 504 norms might therefore guard separately on its own base. It does not; 60 distinct Parameters still give 1 graph.

The 6 in row three is the irreducible set stated in the description, 3 dtypes by 2 grad modes. So in isolation this frame cannot reach 1024, and the guard-axis reduction was never going to be sufficient on its own. That is consistent with the T4 result and makes the earlier claim in this PR wrong for a more specific reason than "it did not work".

What that leaves is the interaction with activation checkpointing, which is where the T4 traceback lands:

```
torch/utils/checkpoint.py in pack_hook
-> target_frame.recomputed[gid][holder.handles[gid]] = x
torch/_dynamo/convert_frame.py in _compile
-> raise FailOnRecompileLimitHit("Hard failure due to fullgraph=True")
```

Under `use_reentrant=False` the region runs twice per step, once packed and once recomputed, and the compile that exhausts the budget happens inside the recompute. I do not yet have the per-guard breakdown from the T4 itself: I re-ran the notebook with `TORCH_LOGS=recompiles` and the env reached the kernel, confirmed by the artifact, but torch's recompile logging did not survive into the captured notebook output, so that run produced no extra signal.

So the open question is narrower than it was: not why a norm recompiles, but why the checkpoint recompute path keeps producing new entries for a frame that is bounded at 6 everywhere I can measure it. I would rather leave this PR judged on what it does measurably do, the width, rank and with_scale axes and bitwise-identical output, than keep it attached to a failure it does not fix.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
One more measurement, and it moves the open question off this kernel entirely.

I reproduced the whole notebook off-T4, on its own stack (torch 2.11.0+cu128, transformers 5.5.0, trl 1.9.2, unsloth 2026.8.5), with `recompile_limit` raised to 8192 so nothing could be truncated. Same model, same LoRA config, same collator, same `use_gradient_checkpointing = "unsloth"`, same `max_length`.

It trains:

```
{'loss': '0.8165', ...}  step 1
{'loss': '0.2275', ...}  step 7
stats: {'calls_captured': 747, 'unique_graphs': 30}
frames: {'total': 11, 'ok': 11}
```

30 unique graphs against a 1024-entry budget. Pinning `dtype = torch.float16` gives 34. Additionally spoofing `torch.cuda.get_device_capability` to `(7, 5)` and `is_bf16_supported` to `False`, so every capability gate in unsloth and unsloth_zoo takes the T4 branch, also gives 34.

So it is neither the compute dtype nor the capability-gated choice of patches. Together with the earlier bounding work (the kernel saturates at 6 unique graphs across width, rank, dtype, grad mode, 60 distinct Parameters, and checkpointed varying shapes), the guard multiplication is not coming from this kernel or from the arithmetic path those gates select.

That leaves this PR standing on what it actually does, which is remove the parameter-width, rank and `with_scale` axes from the compiled frame, forward and backward bitwise identical. It is not the fix for the notebook, and the PR body should not be read as claiming it is. I have a T4 run going with `TORCH_LOGS=recompiles` to get the breakdown; I will post it here either way.

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 1]
```diff
@@ -278,6 +285,34 @@ def forward(self, input_ids: torch.Tensor):
...
+# Clamp bound so a large residual never becomes inf on the cast back. Module-level
+# float keeps `torch.finfo` out of the traced graph.
+_GEMMA4_FP16_MAX = float(torch.finfo(torch.float16).max)
+
+
+# gemma-4-E2B has 504 RMSNorm instances sharing this one code object, so its Dynamo
+# cache holds the product of every axis its guards see. A pure tensor kernel instead
+# of a bound method drops three: `with_scale` (own kernel), parameter width (plain
+# Tensor view) and input rank (2D). Only dtype, grad mode and requires_grad remain.
+@torch_compile(fullgraph = True, dynamic = True)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py]: This is the important one, and it is right. `_fall_back_to_eager_on_recompile_limit` is applied only inside `patch_function`, so a bare `@torch_compile(fullgraph = True)` on a module-level kernel skips it entirely. All five kernels were in that state: `_gemma3_rms_norm_float32`, `_gemma3_rms_norm_generic`, `_gemma4_rms_norm_scaled`, `_gemma4_rms_norm_unscaled`, `_qwen3_moe_rms_norm`.

Fixed in `d70b7691` with a shared `compile_with_eager_fallback(func, label)` in `temporary_patches/utils.py`, applied to all five.

Control on a B200, limit 1, varying dtype and grad mode (widths do not exhaust anything under `dynamic = True`):

```
BARE    -> RAISED FailOnRecompileLimitHit
WRAPPED -> completed
state: {'warned': True, 'eager': True}
```

Worth noting for anyone reading this later: the control deliberately does not set `fail_on_recompile_limit_hit`. That flag is the user asking for a hard stop and the wrapper re-raises for it on purpose; unsloth_zoo leaves it commented out (`patching_utils.py:210`), so a real run is the configuration above. Which means this may well be the fix for the Gemma4 notebook itself, not just a robustness repair. I will say so only once a T4 run shows it.

Tests: 49 passed across this file plus `test_recompile_limit_fallback.py`. The new assertion checks `_unsloth_fallback_state` rather than the absence of `get_compiler_config`, since the wrapper deliberately re-exposes the latter.

File: tests/test_rmsnorm_recompile_guards.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+
+def test_the_old_bound_method_exhausts_a_realistic_recompile_budget():
+    # The regression this fix exists for. The low limit stands in for the real one:
+    # the old form's cache grows with the product of the axes, so any fixed budget
+    # is eventually spent on a model with hundreds of norms.
+    torch._dynamo.reset()
+    with torch._dynamo.config.patch(
+        recompile_limit = 8,
+        fail_on_recompile_limit_hit = True,
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_rmsnorm_recompile_guards.py]: Correct. Fixed in `d70b7691`, following `tests/test_recompile_limit_fallback.py`: the config keys and the exception class are now selected by what the installed torch actually exposes (`recompile_limit`/`fail_on_recompile_limit_hit` vs `cache_size_limit`/`fail_on_cache_limit_hit`), with the file skipping outright when neither a limit exception nor the fail flag exists.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: tests/test_rmsnorm_recompile_guards.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        for dtype in DTYPES:
+            for grad in (True, False):
+                with torch.set_grad_enabled(grad):
+                    x = torch.randn(4, 128, device = "cuda", dtype = dtype)
+                    w = torch.randn(128, device = "cuda", dtype = dtype)
+                    out = _gemma4_rms_norm_scaled(x, w, 1e-6)
+                    assert out.shape == x.shape
+                    assert torch.isfinite(out).all()
+
+    assert _gemma4_rms_norm_scaled._unsloth_fallback_state["eager"], (
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_rmsnorm_recompile_guards.py]: Correct, and it is the same class of leak as the one #1006 exists to fix, so it deserved the same treatment. `torch._dynamo.reset()` does not touch that dict, so every later test in the worker would have used the eager kernel and passed compile assertions without compiling.

Fixed in the latest commit: the states of all five wrappers are saved up front and restored in a `finally`, with a `torch._dynamo.reset()` after, matching the neighbouring test. 26 passed, and the file is order-independent now.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gemma4_float32.py [hunk 4]
```diff
@@ -286,19 +324,30 @@ def patch_Gemma4RMSNorm():
...
+            )
+        else:
+            normed = _gemma4_rms_norm_unscaled(hidden_states_2d, self.eps)
+        return normed.reshape(shape)
     pass
-    patch_function(transformers.models.gemma4.modeling_gemma4.Gemma4RMSNorm, "forward", forward, fullgraph = True, match_level = "relaxed")
+    # No `fullgraph`: the kernels are the compiled units; compiling this wrapper too
+    # would put `self` back into the guards. Dynamo inlines it anyway when a caller is
+    # compiled.
+    patch_function(transformers.models.gemma4.modeling_gemma4.Gemma4RMSNorm, "forward", forward, match_level = "relaxed")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gemma4_float32.py:350]: Measured this on both models rather than reasoning about it. The emitted copy exists, but nothing binds it, so the guards are not restored.

Loaded `unsloth/gemma-4-E2B-it` on this branch through the normal loader (transformers 5.5.0, compute capability spoofed to 7.5 so the float32 norm patches engage, cache deleted first so it really regenerates):

```
== Gemma4RMSNorm ==
  repr        : <function patch_Gemma4RMSNorm.<locals>.forward at 0x722d229c1580>
  __module__  : unsloth_zoo.temporary_patches.gemma4_float32
  is compiled : False

== norm types live in the loaded model ==
  Gemma4RMSNorm                from transformers.models.gemma4.modeling_gemma4
     forward: <function patch_Gemma4RMSNorm.<locals>.forward at 0x722d229c1580>
     compiled: False
```

Same for `unsloth/gemma-3-270m-it` and `Gemma3RMSNorm`. The class the model actually instantiates is still the upstream one carrying this PR's plain wrapper, and the object has no `_torchdynamo_orig_callable` or `get_compiler_config`, so `self.weight`, `self.eps` and the input rank are read in eager exactly as intended.

You are right that the source is re-emitted. `create_standalone_class` reads the live attribute with `inspect.getsource`, so the generated module does contain

```python
@torch.compile(fullgraph = True, dynamic = True, options = torch_compile_options)
def Gemma4RMSNorm_forward(self, x):
```

with this PR's body. But it is never wired up: the generated `Gemma4RMSNorm` class keeps the upstream body and no class delegates to `Gemma4RMSNorm_forward`, unlike siblings such as `Gemma3RotaryEmbedding_forward`, where the class really does `return Gemma3RotaryEmbedding_forward(self, ...)`. I grepped the generated file for a call site and there is none.

Independent corroboration from the failing model itself. On a real T4 the Gemma4 notebook without this PR reports

```
Unsloth: torch.compile ran out of recompilation cache for Gemma4RMSNorm.forward;
finishing this step compiled and switching to eager at the next step.
```

That label and that fallback only come from `patch_function`, so the unit exhausting the budget is the wrapper `patch_function` compiled, not a generated copy. A generated `@torch.compile` has neither, and would have raised instead.

Not disputing that an emitted-but-unbound fullgraph copy is worth being uneasy about. If anything later starts binding those, this PR's premise goes with it.
